### ISSUE TITLE
generated DRR from pelvis CT

### ISSUE THREAD

**Issue by schmidje** [mathiasunberath/DeepDRR#3]
Hi,

Thanks for sharing such a framework, the paper looked very promising!

Like issue #2, I am facing the difficulty to generate a DRR that is similar to a real radiograph. I am trying to replicate some of your results by using an abdominal pelvis CT scan (0.9x0.9x0.4 mm) and by trying to create AP radiograph of the pelvis. I use the same spectrum (120kV, Al 4.3mm) and photons count of 5.10^5 with noise and scatter turned on, but I get a DRR image where areas with small density (air, zones outside of the patient) will present extreme values compared to the body areas of soft or bone. This makes an intensity windowing tedious and not optimal as exemplified in the following picture.

![image](https://user-images.githubusercontent.com/4629132/52209295-c7d59180-2883-11e9-9e3a-4068da3c8e78.png)

I wonder if your images in the paper underwent a last intensity conversion. For instance if I use  logarithmic conversion (-log(DRR+min(DRR))) I get something like:

![image](https://user-images.githubusercontent.com/4629132/52209709-028bf980-2885-11e9-9cc1-c447ee67ff0d.png)

Which has a better dynamics (intensity windowing is easier) but it is not yet appropriate. If I turn off scatter, the image intensities are distributed and the result looks better but still not as nice as in your paper:

![image](https://user-images.githubusercontent.com/4629132/52210081-3c113480-2886-11e9-8e5a-0c7d1fcc51af.png)

Moreover, scatter is turned off which is annoying (and reduces the advantage of using your Deep Learning based framework) and this logarithmic response is somehow arbitrary...

Any pointers?

Thanks!

Best regards,

Jerome

**Comment by mathiasunberath** [comment]
Hi Jerome, 

thank you for your interest in the framework. When you mention that "you wonder whether images in the paper underwent an intensity correction", it would be good if you could point out which Figure exactly you are referring to? The reason I am asking is because, as per its caption, the images shown in our paper (Fig. 3) are in fact **real** X-ray images acquired using a Simenes Cios Fusion C-arm system with flat panel detector. Exemplary images that were generated with DeepDRR are included in the readme, and some examples of log-domain transformed images can be seen here: https://arxiv.org/pdf/1901.06672.pdf.

Some additional comments:
- The in-plane resolution of the CT scans used is fairly low with 0.9mm. This is notable in the images you show, but is not your main concern for now if I understand correctly.
- Images in the DeepDRR framework are output in intensity domain, e.g. highly attenuating structures are dark. This is what a C-arm imaging system would provide. However, it is very true that these images are difficult to window/level due to the high difference in attenuation of bone/air. This is indeed a problem you will have in practice with real systems. 
- Converting to line-integral domain via log transform is encouraged. In fact, when we train CNNs on DeepDRRs we always train on log-transformed images since the dynamic range is easier to handle. 
- Scatter is a complicated topic. You want your generated images to be as close to real X-ray as you can get. Images that you get from a real scanner will have undergone a vendor specific and proprietary pre-processing pipeline that also consists of scatter reduction/removal. E.g. the dynamic density optimization as it is applied in Siemens products. This means that the real X-ray images you target will likely not have scatter, which is why for training set generation you may want to include images with both scatter turned on and off. This is briefly mentioned in the read-me. 
- The benefit of DeepDRR is not only in the integration of deep learning based method, it is one of the few, if not the only, simulation pipeline that allows fast simulation of material and energy dependent X-rays from voxelized representations with compromises regarding the accuracy of scatter. 
- To clarify: There is no heuristic log transform applied anywhere. The log transform can be used to transform from intensity to line-integral domain as per the Beer-Lambert law.

I hope this helps. 
Please let us know if you have further questions or whether this issue can be closed. 
Kind regards,
Mathias

**Comment by schmidje** [comment]
Hi Mathias,

Thanks for such a detailed answer. 

I was referring to the first image of your "representative results" in the README which looks much more contrasted than mine. A higher resolution CT would probably make the image with more details but I don't think it will have a major impact on the image dynamics.

Regarding the log conversion, I checked and you are right it is a common conversion, like in Winslow et al. [1], maybe you could mention that on the documentation. 

Of course your framework does much than a simple DRR generation as it account for the X-ray spectrum. (by the way, could you point me to a place where I could find more spectra to generate similar table like yours?) . However, I really want to include some scatter : ) I wonder about your scattering: does it model/consider some kind of anti scatter grid? If not, I could apply some reduction of the scatter like in [1].

Thanks.

[1] https://www.ncbi.nlm.nih.gov/pubmed/15899304

**Comment by mathiasunberath** [comment]
Hi Jerome.

Another pointer I can give you regarding the appearance of the images is the following: Have you checked that the material decomposition is as you would expect? As you may know from the paper, the current V-net used for material decomposition was trained on a fairly small database since manual labeling was required. As a consequence, it works reasonably well for similar CT scans but we did observe compromised performance when different-looking CT scans were used (this unfortunately can happen easily due to different reconstruction filters, etc). If the V-net does not give you good performance, you can always go back and adjust the thresholds of threshold-based material decomposition (also implemented in DeepDRR) so that it best fits your CT volume (or even provide manual segmentations). 

Regarding scatter: I understand the need for scatter, that is why we tried to also include it. In comparison to other work on machine learning based scatter estimation, we go the "harder way" i.e. trying to estimate what scatter there would be from a scatter-free image. This is more complicated than scatter correction, because scatter is known to be low frequency: If there is scatter signal in the image, a simple low pass filter will already do quite well. In our direction this will not work though, because this low pass signal that we would like to have is not yet in the image. This was just a handwaving argument why our scatter estimation may not perform to your needs. 
The scatter as we model it does not consider any antiscatter grid. If you need reliable scatter signals, I would suggest you rely on low-resolution, variance reduced, forced detection Monte Carlo (e.g. as in [1]) which can be based upon our material decomposition. 
Regarding X-ray spectra: See link in [2], it also allows to model filters. 

Let me know if this answers your questions.
m

[1] https://iopscience.iop.org/article/10.1088/0031-9155/60/4/1415/meta
[2] https://www.oem-xray-components.siemens.com/x-ray-spectra-simulation

**Comment by schmidje** [comment]
Hi,

Once again thanks for the care taken in answering. 

Regarding the V-net segmentation it looked okayish and it can be surely improved I will take care of it, thanks for the pointer.

Regarding the scatter, I can thus anticipate a reduction of it in real radiographs (either via some anti-scatter grid or some advanced post-processing) thus I can reduce the influence of your computed scatter which should improve the visual fidelity while avoiding the first-mentioned issues of difficult intensity range. I am not looking for extremely high fidelity w.r.t the underlying physics but better realism compared to standard simple pass ray tracing for DRR, and your framework has already improved that.

Thanks for the links, [2] is especially very convenient!

I think I can close the "issue" and start experimenting more!

Thanks

**Comment by pranavravella** [comment]
Hello @schmidje! I am currently developing a project that utilizes Bi-planar x-rays from Pelvic CT scans, and the images I got from DeepDRR were all very random for some reason. How far did you get with utilizing DeepDRR for your work and do you have any recommendations for using DeepDRR for pelvic ct scans?

**Comment by mathiasunberath** [comment]
@pranavravella we're happy to help but would require a more detailed description of the problems you are facing.

**Comment by pranavravella** [comment]
Thank you so much for your quick response! One section of my project entails of generating biplanar xrays from pelvic ct scans as aforementioned earlier. However, I'm currently facing two problems.
1) The file at https://www.embodi3d.com/files/file/41843-t12-fx/ generates a resultant image like 
<img width="476" alt="Screen Shot 2020-12-26 at 10 01 55 PM" src="https://user-images.githubusercontent.com/45181376/103163003-06e99480-47c6-11eb-958b-0a949e11bcc0.png">. I was following the example projector in the repo and when printing the materials dictionary, I noticed that it only had True values for air projection and False for other materials. My code is identical to the example projecter btw. 
2) I noticed in the repository you provide phi and theta values. If i wanted to generate the biplanar xrays for a lateral and PA view, how would I modify those values. 
Sorry for the obscurity, I don't know too much about radiography and orthopedics. Thank you for all your help in advance.

**Comment by schmidje** [comment]
Hi,

Various things to consider:
1) properly define your projection and transformation parameters. Projection parameters model the intrinsics of the projection "camera pinhole model" of the X-ray generation process expressed as a K matrix, while transformation parameters express how your CT model is positioned and oriented with respect to your x-ray source ("camera eye") and detector ("image plane"), i.e. a rotation matrix R and a translation vector t. Based on K, R and t you can compute a projection matrix computed by DeepDRR projection matrix module: `proj_mat = projection_matrix.ProjMatrix(R,K,t)`. DeepDRR code will then use this projection matrix in the projector:

```
proj_mats = []
proj_mats.append(proj_mat)

#forward project densities
forward_projections = projector.generate_projections(proj_mats, ....)
```

2) DeepDRR does not model the behavior of a detector in terms of response. Typically radiography has some kind of log conversion (see previous comments in this thread) and you should do the same if you want nicer intensities. In my case for instance I model a 12 bits image as follow:

```
images, photon_prob = mass_attenuation.calculate_intensity_from_spectrum(....)

if logarithmic_response:
	      max_intensity = 2**12
	      images = log_negative_response(images, max_intensity)


def log_negative_response(images, max_intensity_value):
	    
	  for i in range(0, images.shape[0]):
	    image = images[i,:,:]
	    
	    # shift image to avoid invalid values
	    minv = np.min(image)
	    epsilon = 0.000001
	    if minv < epsilon: image += epsilon + minv
	    
	    # take log response
	    image = np.log(image)
	    
	    # negate it and shift it so that log_max_signal_value
	    # is the min
	    image = np.negative(image)
	    
	    min_log_negated_image = np.min(image)
	    max_log_negated_image = np.max(image)
	        
	    # rescale it to be in image dynamic
	    log_negated_image_range = max_log_negated_image-min_log_negated_image
	    images[i,:,:] = (image - min_log_negated_image) * max_intensity_value / log_negated_image_range
	  
	  return images

```

3) Regardless of 2), images may need a final intensity windowing, which should be the responsibility of the user and which is performed via image viewer controls. You can save a preferred windowing in an image format supporting the feature, allowing a viewer to consider it and applying it directly to the image. For instance the medical DICOM image format can do that, although the DICOM format is a little cumbersome to use IMHO.

**Comment by mathiasunberath** [comment]
@schmidje thank you for your detailed response, this is pretty much exactly what I would have answered, too. 

1) Pose parameters: Check that your dicom origin is read properly and that the volume is centered on the pelvis. This is critical to be able to use the diminished theta / phi coordinate system. The other option is to use proper matrices right away. We are working (and have nearly finished) a complete overhaul to the interface that is, in part, designed to make this much easier. @benjamindkilleen has been leading this effort and may be able to comment when it will be available. 

2) Dynamic range is a huge issue in X-ray simulation. Before modifying too much, save your images in a floating point format and inspect them with a viewer made for such images, such as ImageJ as a lightweight option. Use the contrat/brightness or window/level tool to inspect different intensity ranges of the image to check whether information is there (it should be!)

3) The above may still not be enough to get proper looking images after log conversion. We've been working on a detector module that can deal with this, simulating automatic exposure control compensation. @gaocong13, do you have information when we will release this? 

4) Finally, the one image I saw seems to be simulated from a CT scan with low resolution. I would advise to use CT scans with the highest possible resolution.

**Comment by benjamindkilleen** [comment]
@schmidje. To follow up on @mathiasunberath, an updated version of DeepDRR is approaching release. One goal of this revision is to improve the usability and rigorousness of the DeepDRR geometry, as well as to simplify the number of classes/functions the user has to interact with. The update is in the final stages of testing and documentation and will be released in early January.

**Comment by benjamindkilleen** [comment]
@schmidje this has now been released. See branch v1.0.0-alpha.0 or the pre-release, available on PyPI.

**Comment by mathiasunberath** [comment]
@pranavravella see above

**Comment by ackbar03** [comment]
Hi Mathias,

Thanks a lot for the framework and the continued help you've been providing. 

I see in the latest release a inverse log transformation has been added to transform the energy values on the detector to pixel values of an X-ray. Was this the same transformation you used in your research? And also do you happen to know approximately what the real transformation function used in real-life X-ray machines are or where I can find out?

I'm asking because part of my current research work assumes there is roughly a one-to-one relation between the X-ray detector energy and the final x-ray pixel value (i.e. we can infer the relative energy value detected between different places on the same X-ray and also between _different_ X-rays). It would be even better if there is some direct way to transform the pixel values back to detector energy values since my modelling should work better with energy values directly. 

The windowing operation proposed by schmidje for example, if indeed used in real world X-rays, would be a problem for my purposes since different X-rays will be arbitrarily re-scaled using the min-max windowing and the direct one-to-one relationship with absolute energy levels will be lost. 

Do you happen to have any knowledge regarding this or any suggestions? I've been asking around sometime regarding this and also approached an X-ray manufacturer directly, although they seemed a bit reluctant to give a direct answer, possibly because it might be proprietary information, so I didn't press them further. 

Thanks!

**Comment by benjamindkilleen** [comment]
@ackbar03 I can comment on the negative log transform. This is a convenience function not used in prior research, but it may be useful for those just trying to get started.  You can get the original intensities by setting neglog=False in the Projector object.

Regarding the energy values, setting collected_energy=True in the Projector object may be close to what you're looking for.

**Comment by ackbar03** [comment]
@benjamindkilleen  

_This is a convenience function not used in prior research, but it may be useful for those just trying to get started_

Noted!

My question is more about how the intensity/energy levels are converted to actual pixel values stored in a DICOM in real Xray detectors. The framework replicates very closely through simulation of Xray passing through a patient and onto a detector surface. The only part that seems to be missing is how the X-ray intensities / detector energies are converted into the final digital X-ray that is stored in a DICOM that we see in real life. 

Again to further clarify, the rationale of knowing the conversion is to hopefully:
- Increase the fidelity of the DRR simulations further, especially in the transformation from energy/intensity to pixel value
- Allow the possibility of inferring surface detector intensity/energy from real digital X-rays, which after roughly cleaning/accounting for X-ray scatter, allow us to perform rough analysis based on the underlying physics of the ray intensities and energy. 

I've had quite a lot of trouble trying to find clues regarding the conversion function, the closes thing I could find was slide 15 of this obscure document: https://www.aapm.org/meetings/amos2/pdf/29-7999-58461-92.pdf , of which I don't know if it is accurate, up-to-date, over-simplified, or how it should even fit in. Thus I was just wondering if anyone has any insight into this problem.

Thanks again!
